staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go

```diff
@@ -243,142 +296,340 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+			for name, testCase := range testCases {
+				t.Run(name, func(t *testing.T) {
+					server := httptest.NewServer(serverHandler(
+						t, serverHandlerConfig{
+							shouldError:      testCase.shouldError,
+							statusCode:       testCase.serverStatusCode,
+							connectionHeader: testCase.serverConnectionHeader,
+							upgradeHeader:    "SPDY/3.1",
+						},
+					))
+					defer server.Close()
+
+					req, err := http.NewRequest("GET", server.URL, nil)
+					if err != nil {
+						t.Fatalf("error creating request: %s", err)
+					}
+
+					spdyTransport := NewRoundTripper(nil, redirect, redirect)
+
+					var proxierCalled bool
+					var proxyCalledWithHost string
+
+					interceptor := &Interceptor{proxyCalledWithHost: &proxyCalledWithHost}
+
+					proxyHandler := socks5Server(nil, interceptor)
+
+					if testCase.proxyAuth != nil {
+						proxyHandler = socks5Server(&socks5.StaticCredentials{
+							"proxyuser": "proxypasswd",
+						}, interceptor)
+					}
+
+					closed := make(chan struct{})
+					isClosed := func() bool {
+						select {
+						case <-closed:
+							return true
+						default:
+							return false
+						}
+					}
 
-				var proxierCalled bool
-				var proxyCalledWithHost string
-				var proxyCalledWithAuth bool
-				var proxyCalledWithAuthHeader string
-				if testCase.proxyServerFunc != nil {
-					proxyHandler := goproxy.NewProxyHttpServer()
+					l, err := net.Listen("tcp", "127.0.0.1:0")
+					if err != nil {
+						t.Fatalf("socks5Server: proxy_test: Listen: %v", err)
+					}
+					defer func() {
+						close(closed)
+						l.Close()
+					}()
 
-					proxyHandler.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
-						proxyCalledWithHost = host
+					go func(shoulderror bool) {
+						conn, err := l.Accept()
+						if err != nil {
+							if isClosed() {
+								return
+							}
 
-						proxyAuthHeaderName := "Proxy-Authorization"
-						_, proxyCalledWithAuth = ctx.Req.Header[proxyAuthHeaderName]
-						proxyCalledWithAuthHeader = ctx.Req.Header.Get(proxyAuthHeaderName)
-						return goproxy.OkConnect, host
-					})
+							t.Errorf("error accepting connection: %s", err)
+						}
 
-					proxy := testCase.proxyServerFunc(proxyHandler)
+						if err := proxyHandler.ServeConn(conn); err != nil && !shoulderror {
+							if isClosed() {
+								return
+							}
 
+							t.Errorf("ServeConn error: %s", err)
+						}
+					}(testCase.shouldError)
 					spdyTransport.proxier = func(proxierReq *http.Request) (*url.URL, error) {
 						proxierCalled = true
-						proxyURL, err := url.Parse(proxy.URL)
-						if err != nil {
-							return nil, err
+						return &url.URL{
+							Scheme: "socks5",
+							Host:   net.JoinHostPort("127.0.0.1", strconv.Itoa(l.Addr().(*net.TCPAddr).Port)),
+							User:   testCase.proxyAuth,
+						}, nil
+					}
+
+					client := &http.Client{Transport: spdyTransport}
+
+					checkShouldError := func(err error) bool {
```

Comment:
nit: For line 555-576, can we use the same code as line 353-365, which doesn't define a standalone `checkShouldError` function.

Reply:
Agree for both, it should be OK now.